### PR TITLE
Module5:Authorization

### PR DESCRIPTION
**Pull request by derrick1451**
### Requirements
- Install CanCanCan in your project.
- Add a role column to the users table. Remember to use a migration for this.
- A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
- For that you need to implement the post deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.
- A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
- For that you need to implement the comment deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.
### Highlights
RESOLVED

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -18,8 +19,21 @@ def new
...
+    if @post.destroy
+      redirect_to user_posts_path(@user), notice: 'Post deleted successfully!'
+    else
+      redirect_to user_posts_path(@user), alert: 'Error: Post could not be deleted'
+    end
+  end
+
   def create
     @post = current_user.posts.new(params.require(:post).permit(:Title, :Text))
+    authorize! :create, @post
```
Comment (Ibizugbe) [inline app/controllers/posts_controller.rb]: Nice job so far👍🏼. currently, it is impossible to create a post when logged in as a user as it throws the given error below😉. do well to address that as that is the only way we can test for the rest of the required features of the application😉.

image

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @derrick1451 👋🏼👋🏼👋🏼, 

require_changes


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights✅✅✅

- [x] Linters are passing👍🏼.
- [x] ability class has been defined👍🏼.


## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions

- [OPTIONAL] I suggest you add a link that enables a user to sign out and leads to the login page. as this improves the UX of your project👍🏼. 
- [OPTIONAL BUT VERY NECESSARY] I noticed from your previous PR this week that you are paired solo this week, I suggest you always upload an evidence on your PR description for every pair programming activity that is submitted👍🏼. This will avoid any form of delay or misunderstanding amongst reviewers.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/posts/show.html.erb [hunk 2]
```diff
@@ -12,4 +12,7 @@
   
    <button><%= link_to "Add a comment", new_user_post_comment_path(current_user, @post),class:'add-comment' %></button>
    <%= button_to 'Like', user_post_likes_path(current_user, @post), method: :post%>
+    <% if can? :destroy, @post %>
+  <button>  <%= link_to 'Delete', user_post_path(current_user, @post),class:'add-comment', data: { turbo_method: :delete, turbo_confirm: 'Are you sure you want to delete this post?' }%></button>
+  <% end %>
 </div>
```
Comment (TedLivist) [inline app/views/posts/show.html.erb]: - [ ] You have done well implementing the delete feature posts, but I cannot find the button for deleting a comment. Kindly ensure owners of comments and admins can see the delete button and can delete it.

  image

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @derrick1451,

✔️ Delete button and feature for posts

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/posts/show.html.erb (hunk 2 above)
Comment (derrick1451) [reply app/views/posts/show.html.erb]: @TedLivist  a question is the comment button supposed to delete a single comment or all comments in a post

Comment (TedLivist) [reply app/views/posts/show.html.erb]: @derrick1451 the button should delete the comment that is being deleted, not all comments. So a single comment.

Comment (TedLivist) [reply app/views/posts/show.html.erb]: So if there are multiple comments on a post, each comment should have its own delete button - of course depending on the user logged in.

**Review by TadesseDev — APPROVED** [review]
### Hi @derrick1451 

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
AmazingCreepyFrogmouth-max-1mb




## Highlights
 - [x] Good job making the changes requested by the previous code reviewer. ✅

## Optional suggestions

- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**